docs/protocol/src/crypto/tct.md

```diff
@@ -0,0 +1,383 @@
+
+### Structural Definitions
+
+#### Tiers
+
+A tier is either _empty_ or it is a [_node_](#nodes).
+
+#### Nodes
+
+A node in a tree contains a _non-empty_ list of up to four _children_.
+
+If a node has an $n$th child (for $n > 0$), it must have an $(n - 1)$th child (there cannot be gaps
+in the list of children).
+
+Each node must be _left-to-right filled_:
+
+- if its children are nodes, all its children but its rightmost child are _perfect_, and its
+  rightmost child is left-to-right filled (where a _perfect_ node has exactly four children and if its
+  children are nodes, they are also perfect) or
+- if its children are leaves or tiers, the node is left-to-right filled.
+
+#### Leaves
+
+A leaf of a tree is a single commitment.
+
+## Construction
+
+As noted above, the history of all commitments at any point in time is a triply-nested sequence of
+commitments, representing epochs, each of which has some number of blocks, each of which has
+some number of commitments, each ordered at every level chronologically. We can construct a tiered
+commitment tree from this triply nested sequence of commitments by iterating an algorithm to
+construct a single tier from a single sequence of items three times, to reduce a triply-nested
+sequence of commitments to a triply-nested tiered tree.
+
+We define a function $\mathtt{tier}$ which converts a sequence of items (of length less than $4^8$)
+to a tier whose bottom-most nodes' children are those items:
+
+$$
+\begin{align}
+\mathtt{tier}([]) & = \mathtt{Tier}(\emptyset) \\
+\mathtt{tier}(items) & =
+\mathtt{Tier}\big(\underset{1}{\mathtt{chunk}}(\underset{2}{\mathtt{chunk}}(\underset{3}{\mathtt{chunk}}(\underset{4}{\mathtt{chunk}}(\underset{5}{\mathtt{chunk}}(\underset{6}{\mathtt{chunk}}(\underset{7}{\mathtt{chunk}}(\underset{8}{\mathtt{chunk}}(items))))))))[0]\big)
+\end{align}
+$$
+
+where the function $\mathtt{chunk}$ is defined on non-empty sequences as:
+
+$$
+\begin{align}
+\mathtt{chunk}([a, b, c, d, \dots rest]) & = [\mathtt{Node}(a, b, c, d), \dots \mathtt{chunk}(rest)] \\
+\mathtt{chunk}([a, b, c]) & = [\mathtt{Node}(a, b, c)] \\
+\mathtt{chunk}([a, b]) & = [\mathtt{Node}(a, b)] \\
+\mathtt{chunk}([a]) & = [\mathtt{Node}(a)] \\
+\end{align}
+$$
+
+In English, this algorithm is:
+
+- if the sequence is empty, return the empty tier;
+- otherwise, do the following exactly 8 times:
+  - in left to right order, greedily partition the sequence into chunks of at most 4 items, so
+    that all but the last chunk has exactly 4 items, and the last chunk has at most 4 items;
+  - for each chunk, construct a single node from it; and
+  - unless done with the 8 iterations, repeat again, starting with this sequence of nodes (which is
+    now shorter by about a factor of 4)
+- if the input sequence was less than $4^8$ items long, the resultant sequence will now contain
+  exactly one node; return it as the top-most node contained in a tier.
```

Comment:
I'm not sure this adds much over the formulas; I think we just need one definition of the tree structure.